src/ucs/datastruct/ptr_array.h

```diff
@@ -184,6 +201,44 @@ ucs_ptr_array_is_empty(ucs_ptr_array_t *ptr_array)
 }
 
 
+/**
+ * Shift an input value so it will fit in a ptr array slot.
+ *
+ * @param [in] new_val        Value to be put into a slot (at some point).
+ *
+ * @return The same value, as it would be stored in the ptr array.
+ */
+static UCS_F_ALWAYS_INLINE uintptr_t ucs_ptr_array_shift(uintptr_t new_val)
+{
+    return new_val << UCS_PTR_ARRAY_NEXT_SHIFT;
+}
+
+
+/**
+ * Shift back the value retrieved from a ptr array (@ref ucs_ptr_array_shift).
+ *
+ * @param [in] old_val        Value which has been stored in the ptr array slot.
+ *
+ * @return The original value stored in the ptr array slot (before shifting).
+ */
+static UCS_F_ALWAYS_INLINE uintptr_t ucs_ptr_array_unshift(uintptr_t old_val)
+{
+    return old_val >> UCS_PTR_ARRAY_NEXT_SHIFT;
+}
+
+
+/**
+ * Access the element without checking for array bounds or initialization.
+ *
+ * @param [in]  _ptr_array  Pointer to a ptr array.
+ * @param [in]  _index      Index to retrieve the value from.
+ *
+ * @return The "raw" value at the given index, regardless of status.
+ */
+#define __ucs_ptr_array_lookup_no_check(_ptr_array, _index) \
```

Comment:
why need this wrapper?

Reply:
Same: used in #6783 to prevent multiple redundant checks possibly resulting from calling the normal lookup.

Reply:
IMO it would make sense to create a wrapper that does some extra check, why need a wrapper which does nothing?

Reply:
Because when you do multiple consecutive lookups it doesn't make sense to do the check every time (and the compiler might not optimize it away). To clarify: the check we're talking about is for out-of-bounds.

Reply:
if it's not checking anything, why need this macro instead of accessing the array directly?

Reply:
@alex--m this is the only remaining comment I have for this PR

Reply:
@yosefe would you be ok with a function outside ptr_array.[ch] explicitly dereference ucs_ptr_array_t*, i.e. call ptr_array->start[index], like this macro does? If yes - we don't need this macro and I'll remove it.

Reply:
if we need to access elements outside of ptr_array, it better be done by a wrapper (preferably, an inline function)
however, seems this macro is internal (it starts with `__` and is used only by ucs_ptr_array_lookup)
is there another place outside of ptr_array.[ch] where we want to use it?

Reply:
@yosefe Well, yes, we should probably rename it if you agree to the usage, demonstrated here: https://github.com/openucx/ucx/pull/6783/files#r680479980

How about simply `ucs_ptr_array_lookup`?